rest/counted_response_writer.go

```diff
@@ -0,0 +1,61 @@
+// Copyright 2023-Present Couchbase, Inc.
+//
+// Use of this software is governed by the Business Source License included
+// in the file licenses/BSL-Couchbase.txt.  As of the Change Date specified
+// in that file, in accordance with the Business Source License, use of this
+// software will be governed by the Apache License, Version 2.0, included in
+// the file licenses/APL2.txt.
+
+package rest
+
+import (
+	"bufio"
+	"errors"
+	"net"
+	"net/http"
+)
+
+// CountableResponseWriter is an interface that any custom http.ResponseWriter used by Sync Gateway handlers need to implement.
+type CountableResponseWriter interface {
+	http.ResponseWriter
+	GetBytesWritten() int64
+}
+
+// http.ResponseWriter wrapper that counts the number of bytes written in a response. This ignores the bytes written in headers.
+type CountedResponseWriter struct {
+	writer       http.ResponseWriter
+	bytesWritten int64
```

Comment:
Would it be more efficient and/or generate less GC to pass the stat into CountedResponseWriter, and just update the stat directly instead of maintaining a separate int64 on the CountedResponseWriter?

Reply:
I sort of did this and not, the problem is we have to create a the response writer before we know if the database exists or not, so we may or may not be doing anything with this value.

Reply:
If this implies that we're associating a CountedResponseWriter with a request even if it's not a db-scoped request, that doesn't seem ideal.  What's the reason it needs to be initialized that early, as opposed to doing it once we've set the db on the handler?